### PR TITLE
Revert "use git instead of patch by default", instead ignore whitespace when applying diff_file

### LINKED ISSUE #12092: Make diff_file resilient to line-ending differences on Windows
**Question**

I have systems that have patch.exe plus git and others which only have git. I additionally have a diff_file which successfully is applied when only git is present but fails on systems which have patch.exe. It appears wrap files are defaulting to using patch.exe if its present but I want the opposite behavior; always use git. How do I tell a wrap file to always apply the diff_file using git?

### PR DESCRIPTION
**Pull request by nirbheek**
Fixes #12092

### CODE REVIEW

File: mesonbuild/wrap/wrap.py [hunk 1]
```diff
@@ -817,17 +817,17 @@ def apply_diff_files(self) -> None:
...
+                # Always pass a POSIX path to patch, because on Windows it's MSYS
+                # Ignore whitespace when applying patches to workaround
+                # line-ending differences
+                cmd = [PATCH, '-l', '-f', '-p1', '-i', str(Path(relpath).as_posix())]
+            elif GIT:
+                # If the `patch` command is not available, fall back to `git
+                # apply`. The `--work-tree` is necessary in case we're inside a
                 # Git repository: by default, Git will try to apply the patch to
                 # the repository root.
                 cmd = [GIT, '--work-tree', '.', 'apply', '-p1', relpath]
```
Comment (xclaesse) [inline mesonbuild/wrap/wrap.py:830]: I would add `--ignore-whitespace` here too for consistency then. Ideally also have a unit test that mix `\r\n`, I should have done one in the original PR tbh.

**Comment by eli-schwartz** [comment]
Should we limit this to when running on Windows? I'm nervous that this could potentially  cause patches to be messed up (there is a reason this option isn't the default behavior!!)

... on the other hand, people should simply not be using patches :P and this is also part of why I was nervous about implementing diff_files in the first place, so maybe it's a good thing if people get into trouble when they use it?

(I'm only half joking here.)

---

In general I think the reasoning here is good. We cannot automatically assume any form of patch or git-apply or indeed any technology whatsoever in the realm of patch files to ever *automatically* work correctly on Windows with unpredictable line ending checkouts. We can either standardize absolutely on the absolute use of unix line endings (how?) or expect users to manually patch (but we intentionally added support for doing it anyway) or loosen the rules.

**Comment by nirbheek** [comment]
> Should we limit this to when running on Windows

The same problem can also happen on Linux with \r\n line-ending patches added by a developer on Windows.

> Of the options available, loosening the rules for patch to apply patches seems like the only practical one.

Yeah, it's up to the user to rebase patches as and when required. We can only do something best-effort.

File: mesonbuild/wrap/wrap.py (hunk 1 above)
Comment (nirbheek) [reply mesonbuild/wrap/wrap.py:830]: Yeah, makes sense. I'll add it.

**Pull request merged**